subsys/bluetooth/services/ots/ots.c

```diff
@@ -495,6 +546,65 @@ BT_GATT_SERVICE_INSTANCE_DEFINE(ots_service_list, ots_instances,
 				CONFIG_BT_OTS_MAX_INST_CNT,
 				BT_GATT_OTS_ATTRS);
 
+static void ots_delete_empty_name_objects(struct bt_ots *ots)
+{
+	char id_str[BT_OTS_OBJ_ID_STR_LEN];
+	struct bt_gatt_ots_object *obj;
+	struct bt_gatt_ots_object *next_obj;
+	int err;
+
+	err = bt_gatt_ots_obj_manager_first_obj_get(ots->obj_manager, &next_obj);
+	while (!err) {
+		obj = next_obj;
+
+		/* Get the next object before we potentially delete the current object and
+		 * no longer can get the next object
+		 */
+		err = bt_gatt_ots_obj_manager_next_obj_get(ots->obj_manager, obj, &next_obj);
+
+		if (strlen(obj->metadata.name) == 0) {
+			bt_ots_obj_id_to_str(obj->id, id_str, sizeof(id_str));
+			LOG_DBG("Deleting object with %s ID due to empty name", log_strdup(id_str));
+
+			if (ots->cb->obj_deleted) {
+				ots->cb->obj_deleted(ots, NULL, obj->id);
+			}
+
+			if (IS_ENABLED(CONFIG_BT_OTS_DIR_LIST_OBJ)) {
+				bt_ots_dir_list_obj_remove(ots->dir_list, ots->obj_manager,
+							   ots->cur_obj, obj);
+			}
+			bt_gatt_ots_obj_manager_obj_delete(obj);
+		}
+	}
+}
+
+static void ots_conn_disconnected(struct bt_conn *conn, uint8_t reason)
+{
+	uint32_t index;
+	struct bt_ots *instance;
+
+	for (instance = BT_GATT_OTS_INSTANCE_LIST_START, index = 0;
+	     index < instance_cnt;
+	     instance++, index++) {
+
+		LOG_DBG("Processing disconnect for OTS instance %u", index);
+
+		if (instance->cur_obj != NULL) {
+			__ASSERT(instance->cur_obj->state.type == BT_GATT_OTS_OBJECT_IDLE_STATE,
+				"The current object is expected to be in idle state as part "
+				"of cleanup of the L2CAP channel connection close.");
```

Comment:
Will L2CAP always call an callback that will clean up this, before we get the ACL disconnect callback?

Reply:
I believe so. The `bt_l2cap_disconnected` function is called just before the `notify_disconnected` function is called in `deferred_work` function of `subsys/bluetooth/host/conn.c`. This should ensure that all L2CAP channels are closed prior to the disconnect notification to OTS.